sentinel/src/transitions/muting.js

```diff
@@ -0,0 +1,215 @@
+
+const TRANSITION_NAME = 'muting',
+      CONFIG_NAME = 'muting',
+      MUTE_PROPERTY = 'mute_forms',
+      UNMUTE_PROPERTY = 'unmute_forms',
+      SUBJECT_PROPERTIES = ['_id', 'patient_id', 'place_id'];
+
+const getConfig = () => {
+  return config.get(CONFIG_NAME) || {};
+};
+
+const isMuteForm = form => {
+  return Boolean(getConfig()[MUTE_PROPERTY].find(muteFormId => utils.isFormCodeSame(form, muteFormId)));
+};
+
+const isUnmuteForm = form => {
+  const unmuteForms = getConfig()[UNMUTE_PROPERTY];
+  return Boolean(unmuteForms && unmuteForms.find(unmuteFormId => utils.isFormCodeSame(form, unmuteFormId)));
+};
+
+const getContact = doc => {
+  const contactId = doc.fields.patient_id || doc.fields.place_id;
+
+  return db.medic
+    .allDocs({ key: contactId })
+    .then(result => {
+      if (!result.rows.length) {
+        return db.medic.query('medic-client/contacts_by_reference', { key: [ 'shortcode', contactId ] });
+      }
+      return result;
+    })
+    .then(result => {
+      if (!result.rows.length) {
+        module.exports._addErr('contact_not_found', doc);
+        throw(new Error('Contact not found'));
+      }
+      return lineage.fetchHydratedDoc(result.rows[0].id);
+    });
+};
+
+const getDescendants = (contactId) => {
+  return db.medic
+    .query('medic/contacts_by_depth', { key: [contactId] })
+    .then(result => result.rows.map(row => row.id));
+};
+
+const updateDataRecord = (dataRecord, muted) => {
+  return muted ? utils.muteScheduledMessages(dataRecord) : utils.unmuteScheduledMessages(dataRecord);
+};
+
+const updateContacts = (contacts, muted) => {
+  contacts = contacts.filter(contact => {
+    if (Boolean(contact.muted) === muted) {
+      return false;
+    }
+    contact.muted = muted;
+    return true;
+  });
+
+  if (!contacts.length) {
+    return;
+  }
+
+  return db.medic.bulkDocs(contacts);
+};
+
+const updateDataRecords = (subjectIds, muted) => {
+  if (!subjectIds.length) {
+    return Promise.resolve([]);
+  }
+
+  let updatedRecords;
+  return utils.getReportsBySubject({ db: db.medic, ids: subjectIds, registrations: true })
+    .then(dataRecords => {
+      updatedRecords = dataRecords.filter(dataRecord => updateDataRecord(dataRecord, muted));
+      if (!updatedRecords.length) {
+        return;
+      }
+
+      return db.medic.bulkDocs(updatedRecords);
+    })
+    .then(() => updatedRecords);
+};
+
+const getEventType = muted => muted ? 'mute' : 'unmute';
+
+const getContactsAndPatientIds = (doc, contact, muted) => {
+  if (Boolean(contact.muted) === muted) {
+    // don't update registrations if contact already has desired state
+    module.exports._addErr(contact.muted ? 'already_muted' : 'already_unmuted', doc);
+    return;
+  }
+
+  let rootContactId;
+  if (muted) {
+    rootContactId = contact._id;
+  } else {
+    let parent = contact;
+    while (parent && parent.muted) {
+      rootContactId = parent._id;
+      parent = parent.parent;
+    }
```

Comment:
Quote from the design spec:

- There is no such thing as an unmuted person in a muted household. Whenever muting is set at the household level, it applies to all individuals within that household. 
- If a person in a muted household is unmuted, the entire household is unmuted
- We can have a muted person in an unmuted household
- Unmuting an individual person in a muted household should also unmute the entire household and all family members in it (because we don’t allow unmuted individuals underneath a muted household). We can be clear about what will happen on the first page of the individual person’s “unmute” action form.

Reply:
Ok perfect! Thanks for clarifying.